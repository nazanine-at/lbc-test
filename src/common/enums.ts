export enum EMessageType {
  PUBLIC = 'PUBLIC',
  PRIVATE = 'PRIVATE',
}
