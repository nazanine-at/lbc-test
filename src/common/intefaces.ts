import { EMessageType } from './enums'

export interface IState {
  messages: IMessage[]
  selectedMessage: IMessage
}

export interface IMessage {
  userId?: string
  subject?: string
  message?: string
  viewType?: EMessageType
}
