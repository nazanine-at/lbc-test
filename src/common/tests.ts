import configureMockStore, { MockStoreCreator, MockStoreEnhanced } from 'redux-mock-store'
import { IState } from './intefaces'

const initStore: MockStoreCreator = configureMockStore()
export function mockStore(initialState: IState): MockStoreEnhanced {
  return initStore(initialState)
}
