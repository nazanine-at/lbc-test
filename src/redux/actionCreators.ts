import { IMessage } from '../common/intefaces'

// export const setMessagesCreator = (messages: IMessage[]) => {
//   return { type: 'messages/setMessages', payload: messages }
// }
//
// export const addMessageCreator = (message: IMessage) => {
//   return {
//     type: 'message/addMessage',
//     payload: message,
//   }
// }
//
// export const selectMessageCreator = (message: IMessage | undefined) => {
//   return { type: 'message/selectMessage', payload: message }
// }
//
// export const unselectMessageCreator = () => {
//   return { type: 'message/unselectMessage', payload: {} }
// }
