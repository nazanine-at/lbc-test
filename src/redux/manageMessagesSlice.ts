/**
 * This is a reducer - a function that takes a current state value and an
 * action object describing "what happened", and returns a new state value.
 * A reducer's function signature is: (state, action) => newState
 *
 * The Redux state should contain only plain JS objects, arrays, and primitives.
 * The root state value is usually an object.  It's important that you should
 * not mutate the state object, but return a new object if the state changes.
 *
 * You can use any conditional logic you want in a reducer. In this example,
 * we use a switch statement, but it's not required.
 */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { IState } from '../common/intefaces'

export const fetchMessages = createAsyncThunk('fetchMessages', async () => {
  const res = await fetch('/messages')
  return res.json()
})

export const initialState: IState = {
  messages: [],
  selectedMessage: {},
}

const manageMessagesSlice = createSlice({
  name: 'manageMessages',
  initialState,
  reducers: {
    addMessage: (state: IState, { payload }): void => {
      state.messages = [...state.messages, payload]
    },
    selectMessage: (state: IState, { payload }): void => {
      state.selectedMessage = payload
    },
    unselectMessage: (state: IState): void => {
      state.selectedMessage = {}
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchMessages.fulfilled, (state, { payload }) => {
      state.messages = payload
    })
    builder.addCase(fetchMessages.rejected, (state) => {
      state.messages = []
    })
  },
})

export const { unselectMessage, selectMessage, addMessage } = manageMessagesSlice.actions
export default manageMessagesSlice.reducer
