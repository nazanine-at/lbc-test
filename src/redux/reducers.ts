/**
 * This is a reducer - a function that takes a current state value and an
 * action object describing "what happened", and returns a new state value.
 * A reducer's function signature is: (state, action) => newState
 *
 * The Redux state should contain only plain JS objects, arrays, and primitives.
 * The root state value is usually an object.  It's important that you should
 * not mutate the state object, but return a new object if the state changes.
 *
 * You can use any conditional logic you want in a reducer. In this example,
 * we use a switch statement, but it's not required.
 */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { IState } from '../common/intefaces'

export const fetchMessages = createAsyncThunk('fetchMessages', async () => {
  const res = await fetch('/messages')
  return res.json()
})

const manageMessagesSlice = createSlice({
  name: 'manageMessages',
  initialState: {
    messages: [],
    selectedMessage: {},
  },
  reducers: {
    // @ts-ignore
    setMessages: (state: IState, { payload }): void => {
      state.messages = payload.messages
    },
    // @ts-ignore
    addMessage: (state: IState, { payload }): void => {
      state.messages = [...state.messages, payload]
    },
    // @ts-ignore
    selectMessage: (state: IState, { payload }): void => {
      state.selectedMessage = payload
    },
    // @ts-ignore
    unselectMessage: (state: IState): void => {
      state.selectedMessage = {}
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchMessages.fulfilled, (state, { payload }) => {
      state.messages = payload
    })
    builder.addCase(fetchMessages.rejected, (state, action) => {
      state.messages = []
    })
  },
})

export const { actions, reducer } = manageMessagesSlice
