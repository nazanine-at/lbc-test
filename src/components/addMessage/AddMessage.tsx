import { FC, useState, FormEvent, ChangeEvent } from 'react'
import { useDispatch } from 'react-redux'
import { addMessage } from '../../redux/manageMessagesSlice'
import { EMessageType } from '../../common/enums'
import './addMessage.css'

const AddMessage: FC = () => {
  const dispatch = useDispatch()

  const [subject, setSubject] = useState('')
  const [message, setMessage] = useState('')
  const [userId, setUserId] = useState('')
  const [viewType, setViewType] = useState(EMessageType.PUBLIC)

  const onChangeUserId = (e: ChangeEvent<HTMLInputElement>) => setUserId(e.target.value)
  const onChangeSubject = (e: ChangeEvent<HTMLInputElement>) => setSubject(e.target.value)
  const onChangeMessage = (e: ChangeEvent<HTMLTextAreaElement>) => setMessage(e.target.value)
  const updateViewType = (e: ChangeEvent<HTMLSelectElement>) => {
    const element = e.target as HTMLSelectElement
    setViewType(element.value as EMessageType)
  }

  const onSubmitMessage = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    dispatch(addMessage({ userId, subject, message, viewType }))
    setUserId('')
    setSubject('')
    setMessage('')
    setViewType(EMessageType.PUBLIC)
  }

  return (
    <div className="addMessage">
      <form className="form" onSubmit={onSubmitMessage}>
        <label>User name</label>
        <input className="input" placeholder="user name" value={userId} onChange={onChangeUserId} />
        <label>subject</label>
        <input className="input" placeholder="subject" value={subject} onChange={onChangeSubject} />
        <label>Message</label>
        <textarea className="input" value={message} onChange={onChangeMessage} />
        <label htmlFor="messageType">
          View Type
          <select
            className="input"
            id="viewType"
            value={viewType}
            onChange={updateViewType}
            onBlur={updateViewType}
          >
            {[EMessageType.PUBLIC, EMessageType.PRIVATE].map((viewType) => (
              <option key={viewType} value={viewType}>
                {viewType}
              </option>
            ))}
          </select>
        </label>

        <button type="submit">send !</button>
      </form>
    </div>
  )
}

export default AddMessage
