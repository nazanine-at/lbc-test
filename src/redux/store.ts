import { configureStore } from '@reduxjs/toolkit'
import manageMessagesSlice from './manageMessagesSlice'

export const store = configureStore({ reducer: manageMessagesSlice })
