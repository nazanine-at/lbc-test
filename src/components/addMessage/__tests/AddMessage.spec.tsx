import React from 'react'
import { render } from '@testing-library/react'
import AddMessage from '../AddMessage'
import { Provider } from 'react-redux'
import { mockStore } from '../../../common/tests'
import { initialState } from '../../../redux/manageMessagesSlice'

test('<AddMessage/>', () => {
  const store = mockStore(initialState)

  const { container } = render(
    <Provider store={store}>
      <AddMessage />
    </Provider>
  )
  expect(container).toMatchSnapshot()
})
