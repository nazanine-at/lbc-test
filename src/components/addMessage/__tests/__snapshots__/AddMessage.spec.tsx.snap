// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddMessage/> 1`] = `
<div>
  <div
    class="addMessage"
  >
    <form
      class="form"
    >
      <label>
        User name
      </label>
      <input
        class="input"
        placeholder="user name"
        value=""
      />
      <label>
        subject
      </label>
      <input
        class="input"
        placeholder="subject"
        value=""
      />
      <label>
        Message
      </label>
      <textarea
        class="input"
      />
      <label
        for="messageType"
      >
        View Type
        <select
          class="input"
          id="viewType"
        >
          <option
            value="PUBLIC"
          >
            PUBLIC
          </option>
          <option
            value="PRIVATE"
          >
            PRIVATE
          </option>
        </select>
      </label>
      <button
        type="submit"
      >
        send !
      </button>
    </form>
  </div>
</div>
`;
